File: wimg-ios/wimg/Services/WidgetDataWriter.swift
import Foundation
import WidgetKit

enum WidgetDataWriter {
    private static let suiteName = "group.com.wimg.app"
    private static let key = "wimg_widget_data"

    static func writeSummary() {
        let cal = Calendar.current
        let now = Date()
        let year = cal.component(.year, from: now)
        let month = cal.component(.month, from: now)

        let summary = LibWimg.getSummary(year: year, month: month)
        let income = summary.income
        let expenses = summary.expenses
        let savingsRate = income > 0 ? Int(((income - expenses) / income) * 100) : 0

        var data: [String: Any] = [
            "available": summary.available,
            "income": income,
            "expenses": expenses,
            "savings_rate": savingsRate,
            "updated_at": Int(now.timeIntervalSince1970),
        ]

        // Recent transactions for large widget
        let transactions = (try? LibWimg.getTransactions()) ?? []
        let recentTx = transactions.sorted { $0.date > $1.date }.prefix(5).map { tx -> [String: Any] in
            ["description": tx.description, "amount": tx.amount, "date": tx.date]
        }
        data["recent_transactions"] = Array(recentTx)

        let recurring = LibWimg.getRecurring()
            .filter { $0.isActive && $0.next_due != nil }
            .sorted { ($0.next_due ?? "") < ($1.next_due ?? "") }
        if let next = recurring.first {
            data["next_recurring_merchant"] = next.merchant
            data["next_recurring_amount"] = next.amount
            data["next_recurring_date"] = next.next_due
        }

        guard let defaults = UserDefaults(suiteName: suiteName),
              let jsonData = try? JSONSerialization.data(withJSONObject: data),
              let jsonString = String(data: jsonData, encoding: .utf8) else { return }
        defaults.set(jsonString, forKey: key)

        WidgetCenter.shared.reloadAllTimelines()
    }
}

File: wimg-android/app/src/main/java/com/wimg/app/services/WidgetDataWriter.kt
package com.wimg.app.services

import android.content.Context
import androidx.glance.appwidget.updateAll
import com.wimg.app.bridge.LibWimg
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.util.Calendar
import kotlin.math.abs

object WidgetDataWriter {
    private const val PREFS = "wimg_widget"
    private const val KEY = "wimg_widget_data"

    fun writeSummary(context: Context) {
        val cal = Calendar.getInstance()
        val year = cal.get(Calendar.YEAR)
        val month = cal.get(Calendar.MONTH) + 1

        val summary = LibWimg.getSummaryFiltered(year, month, null) ?: return
        val income = summary.income
        val expenses = summary.expenses
        val savingsRate = if (income > 0) ((income - expenses) / income * 100).toInt() else 0

        val recurring = LibWimg.getRecurring()
            .filter { it.active != 0 && it.next_due != null }
            .sortedBy { it.next_due }

        val recent = LibWimg.getTransactions()
            .filter { !it.isExcluded }
            .sortedByDescending { it.date }
            .take(5)

        val data = buildJsonObject {
            put("available", summary.available)
            put("income", income)
            put("expenses", expenses)
            put("savings_rate", savingsRate)
            put("updated_at", System.currentTimeMillis() / 1000)
            recurring.firstOrNull()?.let { next ->
                put("next_recurring_merchant", next.merchant)
                put("next_recurring_amount", next.amount)
                put("next_recurring_date", next.next_due)
            }
            put("recent", buildJsonArray {
                recent.forEach { tx ->
                    add(buildJsonObject {
                        put("desc", tx.description)
                        put("amount", tx.amount)
                    })
                }
            })
        }

        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit().putString(KEY, data.toString()).apply()

        // Trigger widget refresh
        CoroutineScope(Dispatchers.IO).launch {
            try {
                com.wimg.app.widget.WimgSmallWidget().updateAll(context)
                com.wimg.app.widget.WimgMediumWidget().updateAll(context)
                com.wimg.app.widget.WimgLargeWidget().updateAll(context)
            } catch (_: Exception) { }
        }
    }
}
